crates/sparrow-runtime/src/execute/output/object_store.rs

```diff
@@ -1,209 +1,263 @@
-use std::fs::File;
-use std::path::PathBuf;
+use std::str::FromStr;
+use std::sync::Arc;
 
 use arrow::datatypes::SchemaRef;
 use arrow::record_batch::RecordBatch;
-use derive_more::Display;
-use error_stack::{IntoReport, IntoReportCompat, Result, ResultExt};
+use error_stack::{IntoReport, ResultExt};
 use futures::stream::BoxStream;
-use sparrow_api::kaskada::v1alpha::{destination, FileType, ObjectStoreDestination};
-use tempfile::NamedTempFile;
-use tokio::time::Instant;
+use futures::StreamExt;
+use object_store::ObjectStore;
+use parquet::arrow::AsyncArrowWriter;
+use sparrow_api::kaskada::v1alpha::destination::Destination;
+use sparrow_api::kaskada::v1alpha::{FileType, ObjectStoreDestination};
+use tokio::io::{AsyncWrite, AsyncWriteExt};
 use uuid::Uuid;
 
-use crate::{
-    execute::progress_reporter::ProgressUpdate,
-    s3::{is_s3_path, S3Helper, S3Object},
-};
+use crate::execute::progress_reporter::ProgressUpdate;
+use crate::stores::{ObjectStoreRegistry, ObjectStoreUrl};
+use crate::UPLOAD_BUFFER_SIZE_IN_BYTES;
 
-#[derive(Debug, Display)]
+const ROWS_PER_FILE: usize = 1_000_000;
+
+#[derive(derive_more::Display, Debug)]
 pub enum Error {
-    MalformedUri,
-    ProgressUpdateFailure,
-    UnspecifiedFormat,
-    LocalWriteFailure,
-    UploadFailure,
-    UnsupportedUri,
-    UnsupportedDestination,
+    #[display(fmt = "failed to report output progress")]
+    ProgressUpdate,
+    #[display(fmt = "invalid destination prefix")]
+    InvalidDestination,
+    #[display(fmt = "internal error")]
+    Internal,
+    #[display(fmt = "error starting multi-part upload for output")]
+    Upload,
+    #[display(fmt = "error writing output")]
+    Write,
 }
 
 impl error_stack::Context for Error {}
 
+#[must_use]
+enum WriterState {
+    /// Open Parquet writer.
+    Parquet {
+        url: ObjectStoreUrl,
+        writer: AsyncArrowWriter<Box<dyn AsyncWrite + Send + Unpin>>,
+    },
+    Csv {
+        url: ObjectStoreUrl,
+        writer: Box<dyn AsyncWrite + Send + Unpin>,
+    },
+}
+
+impl WriterState {
+    async fn open(
+        object_store: &dyn ObjectStore,
+        output_prefix: &ObjectStoreUrl,
+        file_type: FileType,
+        schema: SchemaRef,
+    ) -> error_stack::Result<Self, Error> {
+        let url = output_prefix
+            .join(&format!("{}", Uuid::new_v4().as_hyphenated()))
```

Comment:
There's currently no "order" to the output files like this, is that acceptable for now we think?

Reply:
I think that should be OK. We could add more book-keeping to try to generate them in a sequence, but that would also change the format of the URL. Although you raise a good point that these file names should at least have a suffix (`.parquet` or `.csv`). Thoughts on naming patterns?

@epinzur

Reply:
for query & materialize file output, we specify the path where all files should go.  see: https://buf.build/kaskada/kaskada/docs/main:kaskada.kaskada.v1alpha#kaskada.kaskada.v1alpha.ObjectStoreDestination

I think that inside that path, order would be super helpful.  

Alternatively, for query, the order could be somewhat determined from the streamed responses. 

For materialize, devs could check file create date, but some sequence would be better.

Files should have extensions.

I don't think URL format matters much.